In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('C:/Users/user/Desktop/k10/k10.csv')


In [3]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100

In [4]:
df['ROE (%)'] = (df['Net Income'] / (df['Total Assets'] - df['Total Liabilities'])) * 100

In [5]:
df['ROA (%)'] = (df['Net Income'] / df['Total Assets']) * 100

In [6]:
df['Net Income Variance'] = (df['Cash Flow Operating Activities'] - df['Net Income'] )

In [7]:
df

Company  Year  Total Revenue    Net Income  Total Assets  \
0  Microsoft  2022   1.982700e+11  7.273800e+10  3.648400e+11   
1  Microsoft  2023   2.119150e+11  7.236100e+10  4.119760e+11   
2  Microsoft  2024   2.451220e+11  8.813600e+10  5.121630e+11   
3      Tesla  2021   5.382300e+10  5.644000e+09  6.213100e+10   
4      Tesla  2022   8.146200e+10  1.258700e+10  8.233800e+10   
5      Tesla  2023   9.677300e+10  1.497400e+10  1.066180e+11   
6      Apple  2021   3.658170e+11  9.468000e+10  3.510020e+11   
7      Apple  2022   3.943280e+11  9.980300e+10  3.527550e+11   
8      Apple  2023   3.832850e+11  9.699500e+10  3.525830e+11   

   Total Liabilities  Cash Flow Operating Activities  Revenue Growth (%)  \
0       1.982980e+11                    8.903500e+10                 NaN   
1       2.057530e+11                    8.758200e+10            6.882030   
2       2.436860e+11                    1.185480e+11           15.669962   
3       3.054800e+10                    1.149700e+10                 NaN   
4       3.644000e+10                    1.472400e+10           51.351653   
5       4.300900e+10                    1.325600e+10           18.795267   
6       2.879120e+11                    1.040380e+11                 NaN   
7       3.020830e+11                    1.221510e+11            7.793788   
8       2.904370e+11                    1.105430e+11           -2.800461   

   Net Income Growth (%)     ROE (%)    ROA (%)  Net Income Variance  
0                    NaN   43.675469  19.936959         1.629700e+10  
1              -0.518299   35.088715  17.564373         1.522100e+10  
2              21.800417   32.828138  17.208584         3.041200e+10  
3                    NaN   17.870373   9.084032         5.853000e+09  
4             123.015592   27.423853  15.286988         2.137000e+09  
5              18.964010   23.540694  14.044533        -1.718000e+09  
6                    NaN  150.071327  26.974205         9.358000e+09  
7               5.410858  196.958873  28.292441         2.234800e+10  
8              -2.813543  156.076015  27.509835         1.354800e+10

## Comprehensive Data Analysis Report for Microsoft

### Findings

**Revenue Growth:**

* Microsoft's revenue has consistently increased over the analyzed period.
* The growth rate accelerated from **6.88%** in 2022-2023 to **15.67%** in 2023-2024, indicating strong business performance and market expansion.

**Profitability:**

* Net income experienced a slight decline in 2022-2023 but rebounded significantly in 2023-2024.
* The **21.80%** increase in net income in 2023-2024 is a positive indicator of improved profitability.

**Return on Equity (ROE):**

* ROE has generally declined over the period, suggesting a less efficient use of shareholder equity.
* A potential explanation could be increased investments in growth initiatives or changes in capital structure.

**Return on Assets (ROA):**

* ROA has also declined, indicating a less efficient use of assets to generate profits.
* This might be due to factors such as increased operating costs or less profitable investments.

**Net Income Variance:**

* The variance between cash from operating activities and net income has increased over the period.
* This could suggest that Microsoft is investing more in growth initiatives or experiencing non-cash expenses that are impacting net income but not cash flow.

### Summary and Financial Health Assessment

Based on the analysis, Microsoft has demonstrated strong revenue growth and a significant rebound in net income in recent years. However, the declining ROE and ROA raise concerns about the efficiency of its asset utilization and shareholder value creation.

The increasing variance between cash from operating activities and net income suggests that Microsoft's financial performance is being influenced by factors beyond its core operations. This could be due to investments in future growth, non-cash expenses, or changes in accounting practices.

Overall, while the company's revenue growth and improved profitability are positive indicators, a more detailed analysis of the factors driving the declining ROA and ROE is necessary to fully assess its financial health. Additionally, the increasing variance between cash from operating activities and net income requires further investigation to understand its implications for the company's long-term sustainability.

### Comprehensive Data Analysis Report for Tesla

**Findings**

**Revenue Growth:**

* Tesla's revenue has experienced significant growth over the analyzed period.
* The growth rate increased from **51.35%** in 2021-2022 to **18.80%** in 2022-2023, indicating strong market expansion and product demand.

**Profitability:**

* Net income has also grown substantially, with a remarkable increase of **123.02%** in 2021-2022 followed by a more moderate growth of **18.96%** in 2022-2023.
* This demonstrates Tesla's ability to improve its profitability and scale its operations efficiently.

**Return on Equity (ROE):**

* ROE has increased over the period, suggesting that Tesla is generating higher returns for its shareholders.
* This is a positive indicator of the company's financial performance and growth potential.

**Return on Assets (ROA):**

* ROA has also increased, indicating that Tesla is becoming more efficient in utilizing its assets to generate profits.
* This improvement in ROA is a positive sign for the company's long-term financial health.

**Net Income Variance:**

* The variance between cash from operating activities and net income has fluctuated over the period.
* In 2021 and 2022, cash from operating activities exceeded net income, suggesting that Tesla was investing in growth initiatives or had non-cash expenses that reduced net income.
* In 2023, the variance became negative, indicating that net income exceeded cash flow. This could be due to factors such as increased working capital requirements or changes in accounting practices.

### Summary and Financial Health Assessment

Based on the analysis, Tesla has demonstrated exceptional financial performance, characterized by strong revenue and net income growth, improving profitability metrics (ROE and ROA), and significant investments in growth initiatives.

The fluctuating net income variance highlights the dynamic nature of Tesla's business and its strategic investments. While the company has experienced periods of higher cash outflows, its overall financial health appears to be robust.

**Key Strengths:**

* Strong revenue and net income growth
* Improving profitability metrics (ROE and ROA)
* Strategic investments in growth initiatives

**Areas for Further Analysis:**

* The factors driving the fluctuations in net income variance
* The sustainability of Tesla's high growth rate
* The company's competitive position in the electric vehicle market

Overall, Tesla's financial performance is impressive, and the company appears to be well-positioned for continued success. However, ongoing monitoring and analysis of key financial metrics are essential to ensure its long-term financial health and sustainability.

### Comprehensive Data Analysis Report for Apple

**Findings**

**Revenue Growth:**

* Apple's revenue experienced moderate growth in 2021-2022, but declined in 2022-2023.
* The decline in revenue growth could be attributed to various factors, such as economic conditions, supply chain disruptions, or changes in consumer preferences.

**Profitability:**

* Net income followed a similar trend to revenue, with a slight increase in 2021-2022 followed by a decline in 2022-2023.
* The decline in net income growth is likely due to increased costs, decreased sales, or other factors affecting profitability.

**Return on Equity (ROE):**

* ROE has remained relatively high throughout the analyzed period, indicating efficient use of shareholder equity.
* The slight decline in ROE in 2023 could be attributed to increased investments or changes in capital structure.

**Return on Assets (ROA):**

* ROA has remained relatively stable, suggesting consistent asset utilization efficiency.
* The slight increase in ROA in 2022 followed by a slight decline in 2023 indicates that Apple's asset management practices have been relatively consistent.

**Net Income Variance:**

* The variance between cash from operating activities and net income has increased over the period.
* This suggests that Apple has been investing more in growth initiatives or experiencing non-cash expenses that have impacted net income but not cash flow.

### Summary and Financial Health Assessment

Based on the analysis, Apple has maintained a relatively strong financial position despite a decline in revenue and net income growth in 2022-2023. The company's high ROE and stable ROA indicate efficient use of equity and assets.

The increasing variance between cash from operating activities and net income suggests that Apple has been investing in growth initiatives or experiencing non-cash expenses. While this is a positive sign of future growth potential, it's important to monitor these investments closely to ensure they are generating returns.

Overall, Apple's financial health appears to be solid, but the company may need to address the factors contributing to the decline in revenue and net income growth to maintain its long-term competitiveness and profitability.
